RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc

```diff
@@ -10,44 +11,430 @@
   edm::ESGetToken<TrackerTopology, TrackerTopologyRcd> ttopoToken_;
+  edm::ESGetToken<GeometricSearchTracker, TrackerRecoGeometryRecord> trackerToken_;
+
+  const TrackerTopology *m_trackerTopo = nullptr;
+  const TrackerGeometry *m_trackerGeom = nullptr;
+  mkfit::LayerNumberConverter m_layerNrConv = {mkfit::TkLayout::phase1};
 };
 
-MkFitGeometryESProducer::MkFitGeometryESProducer(const edm::ParameterSet& iConfig) {
+MkFitGeometryESProducer::MkFitGeometryESProducer(const edm::ParameterSet &iConfig) {
   auto cc = setWhatProduced(this);
   geomToken_ = cc.consumes();
-  trackerToken_ = cc.consumes();
   ttopoToken_ = cc.consumes();
+  trackerToken_ = cc.consumes();
 }
 
-void MkFitGeometryESProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+void MkFitGeometryESProducer::fillDescriptions(edm::ConfigurationDescriptions &descriptions) {
   edm::ParameterSetDescription desc;
   descriptions.addWithDefaultLabel(desc);
 }
 
-std::unique_ptr<MkFitGeometry> MkFitGeometryESProducer::produce(const TrackerRecoGeometryRecord& iRecord) {
+//------------------------------------------------------------------------------
+
+void MkFitGeometryESProducer::GapCollector::add_interval(float x, float y) {
+  if (x > y)
+    std::swap(x, y);
+  bool handled = false;
+  for (auto i = m_coverage.begin(); i != m_coverage.end(); ++i) {
+    if (y < i->x) {  // fully on 'left'
+      m_coverage.insert(i, {x, y});
+      handled = true;
+      break;
+    } else if (x > i->y) {  // fully on 'right'
+      continue;
+    } else if (x < i->x) {  // sticking out on 'left'
+      i->x = x;
+      handled = true;
+      break;
+    } else if (y > i->y) {  // sticking out on 'right'
+      i->y = y;
+      // check for overlap with the next interval, potentially merge
+      auto j = i;
+      ++j;
+      if (j != m_coverage.end() && i->y >= j->x) {
+        i->y = j->y;
+        m_coverage.erase(j);
+      }
+      handled = true;
+      break;
+    } else {  // contained in current interval
+      handled = true;
+      break;
+    }
+  }
+  if (!handled) {
+    m_coverage.push_back({x, y});
+  }
+}
+
+void MkFitGeometryESProducer::GapCollector::sqrt_elements() {
+  for (auto &itvl : m_coverage) {
+    itvl.x = std::sqrt(itvl.x);
+    itvl.y = std::sqrt(itvl.y);
+  }
+}
+
+bool MkFitGeometryESProducer::GapCollector::find_gap(Interval &itvl, float eps) {
+  auto i = m_coverage.begin();
+  while (i != m_coverage.end()) {
+    auto j = i;
+    ++j;
+    if (j != m_coverage.end()) {
+      if (j->x - i->y > eps) {
+        itvl = {i->y, j->x};
+        return true;
+      }
+      i = j;
+    } else {
+      break;
+    }
+  }
+  return false;
+}
+
+void MkFitGeometryESProducer::GapCollector::print_gaps() {
+  auto i = m_coverage.begin();
+  while (i != m_coverage.end()) {
+    auto j = i;
+    ++j;
+    if (j != m_coverage.end()) {
+      printf("(%f, %f)->%f ", i->y, j->x, j->x - i->y);
```

Comment:
LogVerbatim perhaps?

Reply:
I can look it up and replace ... but this only called for crosscheck when tuning the ignored gap-size when r-coverage is spotty. I suspect I'll only need it when working on the phaseII geometry, if it also has a coverage gap :)